src/foam/nanos/crunch/ui/CapabilityView.js

```diff
@@ -0,0 +1,111 @@
+foam.CLASS({
+  package: 'foam.nanos.crunch.ui',
+  name: 'CapabilityView',
+  extends: 'foam.u2.View',
+  documentation: 'All purpose capability view that takes in an array of capaiblity ids and displays these capabilities',
+
+  imports: [
+    'crunchController',
+    'notify'
+  ],
+
+  requires: [
+    'foam.u2.detail.SectionView'
+  ],
+
+  properties: [
+    {
+      class: 'StringArray',
+      name: 'capabilities'
+    },
+    {
+      name: 'wizardlets',
+      documentation: 'wizardlets for the capability',
+      value: []
+    },
+    {
+      name: 'wizardletSectionsList',
+      documentation: `
+        sections for wizardlets
+        wizardletSectionsList[i] stores sections for wizardlets[i]
+      `,
+      value: []
+    }
+  ],
+
+  methods: [
+    async function initE() {
+      const self = this;
+
+      // gets the wizardlets and sections for all of the capabilties
+      for ( let cap of this.capabilities ) {
+        // wizardlets for current capability
+        const curWizardlets = (await this.crunchController.getCapsAndWizardlets(cap)).wizCaps;
+
+         // adds listener on curWizardlets
+         for ( let wizardlet of curWizardlets ) {
+          this.addListeners(wizardlet.data);
+        }
+
+        // gets all the sections associated with curWizardlets
+        const curWizardletSectionsList = this.crunchController.generateSections(curWizardlets);
+
+        this.wizardlets = this.wizardlets.concat(curWizardlets);
+        this.wizardletSectionsList = this.wizardletSectionsList.concat(curWizardletSectionsList);
+      }
+
+      this.start().addClass(this.myClass())
+        .add(self.slot(function(wizardletSectionsList) {
+          return this.E().forEach(wizardletSectionsList, function(sections, index) {
+            sections.map(section => (
+              this.tag(self.SectionView, {
+                section,
+                data: self.wizardlets[index].data,
+                showTitle: false
+              })
+            ));
+          });
+        }))
+      .end();
+    },
+
+    // adds a listener to each wizardlet.data obj for the purpose of saving
+    // all user inputs releasing calling views from the responsibility
+    function addListeners(obj) {
+      if ( ! this.isFObject(obj) ) return;
```

Comment:
why not just check for null? You're checking for isFObject below and the first iteration through the function will be an object or null since it references the "of" of a capability.

Reply:
@zahnno Yes, if we call the function inside the loop, it will guarantee that the argument passed to the function will be fobject. But if we call the function for the first time, there is no protection against passing a non fobject to the function. In this case, calling sub on `obj` will raise an error.

Reply:
@zahnno Since we know `wizardlet.data` is an fobject, it will run into the case above, but I still added there for extra layer of protection.